File: tests/fixtures/vulnerable/taint.java
// FIXTURE: verified data flows, Java (Servlet API + JDBC).
import java.sql.*;
import javax.servlet.http.HttpServletRequest;

public class TaintFixture {
    public ResultSet throughAVariable(HttpServletRequest request, Statement stmt) throws SQLException {
        String id = request.getParameter("id");
        String sql = "SELECT * FROM users WHERE id = " + id;
        // EXPECT-FLOW sql-injection
        return stmt.executeQuery(sql);
    }

    public void intoAShell(HttpServletRequest request) throws Exception {
        // EXPECT-FLOW command-injection
        Runtime.getRuntime().exec("ping -c 1 " + request.getParameter("host"));
    }

    private String buildQuery(String value) {
        return "SELECT * FROM users WHERE name = '" + value + "'";
    }

    public ResultSet acrossFunctions(HttpServletRequest request, Statement stmt) throws SQLException {
        // EXPECT-FLOW sql-injection
        return stmt.executeQuery(buildQuery(request.getParameter("q")));
    }
}

File: tests/fixtures/vulnerable/taint.go
// FIXTURE: verified data flows, Go (net/http + database/sql).
package fixtures

import (
	"database/sql"
	"net/http"
	"os/exec"
)

func throughAVariable(r *http.Request, db *sql.DB) {
	id := r.FormValue("id")
	query := "SELECT * FROM users WHERE id = " + id
	// EXPECT-FLOW sql-injection
	db.Query(query)
}

func intoAShell(r *http.Request) error {
	// EXPECT-FLOW command-injection
	return exec.Command("sh", "-c", "ping -c 1 "+r.FormValue("host")).Run()
}

func buildQuery(value string) string {
	return "SELECT * FROM users WHERE name = '" + value + "'"
}

func acrossFunctions(r *http.Request, db *sql.DB) {
	// EXPECT-FLOW sql-injection
	db.Query(buildQuery(r.FormValue("q")))
}
